src/renderers/shared/fiber/ReactFiberScheduler.js

```diff
@@ -267,6 +274,11 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
       const current = workInProgress.alternate;
       const next = completeWork(current, workInProgress);
 
+      // We are leaving this subtree, so pop context if any.
+      if (isContextProvider(workInProgress)) {
+        popContextProvider();
```

Comment:
I was hoping that we would be able to put push/pop inside ReactFiberBeginWork/ReactFiberCompleteWork. The reason is that I'd like to be able to reuse the tag check inside the switch statement so that we only even check for this if we hit the ClassComponent branch. That way when we build optimized paths, they can completely skip this check.

I realize that this might be currently broken because we bail out *before* the switch statement in certain cases right now. We might have to move those around.